### ISSUE TITLE
mac m2 run air llm garage-bAInd/Platypus2-7B  get error  Input must be a file-like object opened in binary mode, or string

### ISSUE THREAD

**Issue by wuxiongwei** [lyogavin/airllm#116]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[23], [line 2](vscode-notebook-cell:?execution_count=23&line=2)
      [1](vscode-notebook-cell:?execution_count=23&line=1) import mlx.core as mx
----> [2](vscode-notebook-cell:?execution_count=23&line=2) generation_output = model.generate(
      [3](vscode-notebook-cell:?execution_count=23&line=3)     mx.array(input_tokens['input_ids']), 
      [4](vscode-notebook-cell:?execution_count=23&line=4)     max_new_tokens=3,
      [5](vscode-notebook-cell:?execution_count=23&line=5)     use_cache=True,
      [6](vscode-notebook-cell:?execution_count=23&line=6)     return_dict_in_generate=True)
      [8](vscode-notebook-cell:?execution_count=23&line=8) print(generation_output)

File [~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:254](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:254), in AirLLMLlamaMlx.generate(self, x, temperature, max_new_tokens, **kwargs)
    [252](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:252) def generate(self, x, temperature=0, max_new_tokens=None, **kwargs):
    [253](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:253)     tokens = []
--> [254](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:254)     for token in self.model_generate(x, temperature=temperature):
    [255](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:255)         tokens.append(token)
    [258](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:258)         if len(tokens) >= max_new_tokens:

File [~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:281](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:281), in AirLLMLlamaMlx.model_generate(self, x, temperature, max_new_tokens)
    [278](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:278) mask = mask.astype(self.tok_embeddings.weight.dtype)
    [280](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:280) self.record_memory('before_loading_tok')
--> [281](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:281) update_weights = ModelPersister.get_model_persister().load_model(self.layer_names_dict['embed'], self.checkpoint_path)
    [283](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:283) self.record_memory('after_loading_tok')
    [284](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/airllm_llama_mlx.py:284) self.tok_embeddings.update(update_weights['tok_embeddings'])
...
     [97](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/persist/mlx_model_persister.py:97) #available = psutil.virtual_memory().available [/](https://file+.vscode-resource.vscode-cdn.net/) 1024 [/](https://file+.vscode-resource.vscode-cdn.net/) 1024
     [98](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/persist/mlx_model_persister.py:98) #print(f"loaded {layer_name}, available mem: {available:.02f}")
    [100](https://file+.vscode-resource.vscode-cdn.net/Users/zoe/Documents/codes/0030python/Anima/air_llm/examples/~/opt/anaconda3/envs/air_llm_python_3_8/lib/python3.8/site-packages/airllm/persist/mlx_model_persister.py:100) layer_state_dict = map_torch_to_mlx(layer_state_dict)

ValueError: [load] Input must be a file-like object opened in binary mode, or string

**Comment by Verdagon** [comment]
I just had this same problem, and I think I have a fix. In `mlx_model_persister.py`, change:
`mx.load(to_load_path)`
to:
`mx.load(str(to_load_path))`

@wuxiongwei, can you help verify the fix works for you?

**Comment by 8-momo-8** [comment]
@Verdagon , yes it works with converting it to string 'mx.load(str(to_load_path))'

**Comment by mustangs0786** [comment]
### Help where i have to put above code.....

input_text = [
        #'What is the capital of United States?',
        'I like',
    ]

MAX_LENGTH = 128
input_tokens = model.tokenizer(input_text,
    return_tensors="np", 
    return_attention_mask=False, 
    truncation=True, 
    max_length=MAX_LENGTH, 
    padding=False)

input_tokens



generation_output = model.generate(
    mx.array(input_tokens['input_ids']), 
    max_new_tokens=3,
    use_cache=True,
    return_dict_in_generate=True)

print(generation_output)

**Comment by shahfasal** [comment]
@mustangs0786 same error for me where you able to figure it out?

**Comment by shahfasal** [comment]
Got it its under airllm/persist/

**Comment by lyogavin** [comment]
Closing as stale. This concerns the Apple-Silicon/MLX backend with old mlx versions; that path is unchanged and these reports are from much older versions. If you still hit an MLX issue on current versions, please open a fresh issue with your mlx/macOS versions and a reproducer.
